src/category_theory/bicategory/coherence.lean

```diff
@@ -0,0 +1,291 @@
+Copyright (c) 2022 Yuma Mizuno. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yuma Mizuno
+-/
+import category_theory.path_category
+import category_theory.fully_faithful
+import category_theory.bicategory.free
+import category_theory.bicategory.locally_discrete
+/-!
+# The coherence theorem for bicategories
+
+In this file, we prove the coherence theorem for bicategories, stated in the following form: the
+free bicategory over any quiver is locally thin.
+
+The proof is almost the same as the proof of the coherence theorem for monoidal categories that
+has been previously formalized in mathlib, which is based on the proof described by Ilya Beylin
+and Peter Dybjer. The idea is to view a path on a quiver as a normal form of a 1-morphism in the
+free bicategory on the same quiver. A normalization procedure is then described by
+`full_normalize : pseudofunctor (free_bicategory B) (locally_discrete (paths B))`, which is a
+pseudofunctor from the free bicategory to the locally discrete bicategory on the path category.
+It turns out that this pseudofunctor is locally an equivalence of categories, and the coherence
+theorem follows immediately from this fact.
+
+## Main statements
+
+* `locally_thin` : the free bicategory is locally thin, that is, there is at most one
+  2-morphism between two fixed 1-morphisms.
+
+## References
+
+* [Ilya Beylin and Peter Dybjer, Extracting a proof of coherence for monoidal categories from a
+   proof of normalization for monoids][beylin1996]
+-/
+
+open quiver (path) quiver.path
+
+namespace category_theory
+
+open bicategory category
+open_locale bicategory
+
+universes v u
+
+namespace free_bicategory
+
+variables {B : Type u} [quiver.{v+1} B]
+
+/-- Auxiliary definition for `inclusion_path`. -/
+@[simp]
+def inclusion_path_aux {a : B} : ∀ {b : B}, path a b → hom a b
+| _ nil := hom.id a
+| _ (cons p f) := (inclusion_path_aux p).comp (hom.of f)
+
+/--
+The discrete category on the paths includes into the category of 1-morphisms in the free
+bicategory.
+-/
+def inclusion_path (a b : B) : discrete (path.{v+1} a b) ⥤ hom a b :=
+{ obj := inclusion_path_aux,
+  map := λ f g η, eq_to_hom (congr_arg inclusion_path_aux (discrete.eq_of_hom η)) }
+
+variables (B)
+
+/--
+The inclusion from the locally discrete bicategory on the path category into the free bicategory
+as a prelax functor. This will be promoted to a pseudofunctor after proving the coherence theorem.
+See `inclusion`.
+-/
+@[simps]
+def preinclusion : prelax_functor (locally_discrete (paths B)) (free_bicategory B) :=
+{ obj := id,
+  map := λ a b, (inclusion_path a b).obj,
+  map₂ := λ a b, (inclusion_path a b).map }
+
+variables {B}
+
+/--
+The normalization of the composition of `p : path a b` and `f : hom b c`. Defining this function
+is easier than defining the normalization of `f : hom a b` alone, which will defined as the
+normalization of the composition of `path.nil : path a a` and `f : hom a b`.
```

Comment:
I had no trouble defining hom a b -> path a b directly in https://github.com/leanprover-community/mathlib/compare/bicategory-free-coherence...bicategory-free-coherence'?expand=1 Does any difficulty arises elsewhere?

And in this PR I don't seem to see any substantial use of the `pseudofunctor` structure. I'd suggest you make `pseudofunctor` extend `oplax_functor`, and if (in the future) `is_iso map_id` and `is_iso map_comp` are insufficient for you and you must use the data of the inverses, you could then add the fields `map_id_inv`, `map_comp_inv` and `inv_comp`, `comp_inv` conditions, as in [groupoids](https://leanprover-community.github.io/mathlib_docs/category_theory/groupoid.html#category_theory.groupoid), which will be shorter than the current definition of pseudofunctor, and no additional constructors would be needed.